Python/ceval.c

```diff
@@ -4603,10 +4621,44 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
 
         TARGET(CALL_FUNCTION): {
             PREDICTED(CALL_FUNCTION);
-            PyObject **sp, *res;
-            sp = stack_pointer;
-            res = call_function(tstate, &sp, oparg, NULL, cframe.use_tracing);
-            stack_pointer = sp;
+            PyObject *res;
+
+            // Check if the call can be inlined or not
+            PyObject *function = PEEK(oparg + 1);
+            if (Py_TYPE(function) == &PyFunction_Type) {
+                PyCodeObject *code = (PyCodeObject*)PyFunction_GET_CODE(function);
+                PyObject *locals = code->co_flags & CO_OPTIMIZED ? NULL : PyFunction_GET_GLOBALS(function);
+                if ((code->co_flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR)) == 0) {
+                    InterpreterFrame *new_frame = _PyEvalFramePushAndInit(
+                        tstate, PyFunction_AS_FRAME_CONSTRUCTOR(function), locals, stack_pointer-oparg, oparg, NULL, 1);
+                    if (new_frame == NULL) {
+                        // When we exit here, we own all variables in the stack (the frame creation has not stolen
+                        // any variable) so we need to clean the whole stack (done in the "error" label).
+                        goto error;
+                    }
+                    STACK_SHRINK(oparg + 1);
+                    assert(tstate->interp->eval_frame != NULL);
+                    // The frame has stolen all the arguments from the stack, so there is no need to clean them up.```
+                    Py_DECREF(function);
+                    _PyFrame_SetStackPointer(frame, stack_pointer);
+                    new_frame->depth = frame->depth + 1;
+                    tstate->frame = frame = new_frame;
+                    goto start_frame;
+                }
+                else {
+                    /* Callable is a generator or coroutine function: create coroutine or generator. */
+                    res = make_coro(tstate, PyFunction_AS_FRAME_CONSTRUCTOR(function), locals, stack_pointer-oparg, oparg, NULL);
+                    STACK_SHRINK(oparg + 1);
+                    for (int i = 0; i < oparg + 1; i++) {
+                        Py_DECREF(stack_pointer[i]);
+                    }
+                }
+            }
+            else {
+                PyObject **sp = stack_pointer;
+                res = call_function(tstate, &sp, oparg, NULL, cframe.use_tracing);
```

Comment:
A curiosity: To reduce stack usage, bpo-29227 marked `call_function()` as `Py_LOCAL_INLINE`. Depending on how often this branch is hit, perhaps it would be worth it to shrink the effective body of `_PyEval_EvalFrameDefault` by un-inlining?

Then again, maybe that would only be worth it if the other usages of `call_function()` (`CALL_METHOD`, `CALL_METHOD_KW`, `CALL_FUNCTION_KW`) could be similarly bypassed.

Reply:
If I understand correctly, I interpreted that if this path is less often hit, it may be better to not inline ``call_function`` to reduce the code size of `_PyEval_EvalFrameDefault`.

IMO, we don't need to care about that because PGO should be smart enough to decide when to inline/not inline. Also from my own profiling, >60% of callables at runtime are C PyCFunction/Method/ClassMethod etc. Python function makes up maybe < 20% of all calls to `CALL_FUNCTION` on average. So this branch will still be quite common and we should still mark ``Py_LOCAL_INLINE``..

Reply:
>  I interpreted that if this path is less often hit, it may be better not to inline call_function to reduce the code size of _PyEval_EvalFrameDefault.

Unfortunately I still don't understand what you mean. "Not inline call_function" is the same as closing this PR, no?

Reply:
Ah, I understand now. Yeah, I agree with has been said: the branch is still common enough even after this change and I don't want to sacrifice performance for the default case as well.